rgsync/Connectors/redis_connector.py

```diff
@@ -0,0 +1,177 @@
+            WriteBehindLog(msg)
+            raise Exception(msg) from None
+        return r
+
+
+# redis cluster connection class
+class RedisClusterConnection():
+    def __init__(self, host=None, port=None, cluster_nodes=None, password=None):
+        self._host = host
+        self._port = port
+        self._cluster_nodes = cluster_nodes
+        self._password = password
+
+    @property
+    def host(self):
+        return self._host() if callable(self._host) else self._host
+
+    @property
+    def port(self):
+        return self._port() if callable(self._port) else self._port
+
+    @property
+    def password(self):
+        return self._password() if callable(self._password) else self._password
+
+    @property
+    def cluster_nodes(self):
+        return self._cluster_nodes() if callable(self._cluster_nodes) else self._cluster_nodes
+    
+    def Connect(self):
+        from rediscluster.client import RedisCluster
+        try:
+            WriteBehindLog("Connect: connecting to {}:{} password: {} cluster nodes: {}".format(self.host, self.port, self.password, self.cluster_nodes))
+            rc = RedisCluster(host=self.host, port=self.port, startup_nodes=self.cluster_nodes, password=self.password, decode_responses=True)
+        except Exception as e:
+            msg = "Cannot connect to Redis Cluster. Exception: {}".format(e)
+            WriteBehindLog(msg)
+            raise Exception(msg) from None
+        return rc
+
+
+SIMPLE_HASH_BACKEND_PK = 'HashBackendPK'
+SIMPLE_HASH_BACKEND_TABLE = 'HashBackendTable'
+
+# redis connector class
+class RedisConnector():
+    def __init__(self, connection, newPrefix, exactlyOnceTableName=None):
+        self.connection = connection
+        self.session = None
+        self.new_prefix = newPrefix
+        self.supportedOperations = [OPERATION_DEL_REPLICATE, OPERATION_UPDATE_REPLICATE]
+        self.exactlyOnceTableName = exactlyOnceTableName
+        self.exactlyOnceLastId = None
+        self.shouldCompareId = True if self.exactlyOnceTableName is not None else False
+
+    def TableName(self):
+        return SIMPLE_HASH_BACKEND_TABLE
+
+    def PrimaryKey(self):
+        return SIMPLE_HASH_BACKEND_PK
+
+    def WriteData(self, data):
+        # return if no data is received to write
+        if(0 == len(data)):
+            WriteBehindLog("Warning: in connector received empty batch to write")
+            return
+
+        if not self.session:
+            self.session = self.connection.Connect()
+        
+        # in case of exactly once get last id written
+        shardId = None
+        try:
+            # Raise eception in case of exactly once property is used in RedisCluster
+            if((self.exactlyOnceTableName is not None) and
+                    isinstance(self.connection, RedisClusterConnection)):
+                msg = "Exactly once property is not valid for Redis cluster"
+                raise Exception(msg) from None
```

Comment:
We can do this test in the __init__ function to raise the error as soon as possible and not let the registration to be registered from the first place.

Reply:
I will update the code